OSDOCS-6582: CLIOT-28 CPU Usage guidance (#82199)
openshift/openshift-docs: modules/admin-quota-overview.adoc

:_mod-docs-content-type: CONCEPT
[id="admin-quota-overview_{context}"]
= Setting Quotas

A resource quota, defined by a `ResourceQuota` object, provides constraints that limit aggregate resource consumption per project. It can limit the quantity
of objects that can be created in a project by type, as well as the total amount of compute resources and storage that may be consumed by resources in that project.

ifdef::openshift-origin,openshift-enterprise,openshift-dedicated[]

[NOTE]
====
See xref:../applications/quotas/quotas-setting-per-project.adoc[Resources managed by quotas] for more on compute resources.
====
endif::[]

[id="resources-managed-by-quota"]
== Resources Managed by Quota

The following describes the set of compute resources and object types that may be managed by a quota.

[NOTE]
====

> A pod is in a terminal state if `status.phase in (Failed, Succeeded)` is true.

Author: @snarayan-redhat THANK YOU this is done!

|`pods`
|The total number of pods in a non-terminal state that can exist in the project.

|`replicationcontrollers`
|The total number of replication controllers that can exist in the project.

|`resourcequotas`
|The total number of resource quotas that can exist in the project.

|`services`
|The total number of services that can exist in the project.

|`secrets`
|The total number of secrets that can exist in the project.

|`configmaps`
|The total number of `ConfigMap` objects that can exist in the project.

|`persistentvolumeclaims`
|The total number of persistent volume claims that can exist in the project.

|`openshift.io/imagestreams`
|The total number of image streams that can exist in the project.

|===
end::admin_quota_resources_managed[]

You can configure an object count quota for these standard namespaced resource types using the `count/<resource>.<group>` syntax.

[source,terminal]
----
$ oc create quota <name> --hard=count/<resource>.<group>=<quota> <1>
----

<1> `<resource>` is the name of the resource, and `<group>` is the API group, if applicable.
Use the `kubectl api-resources` command for a list of resources and their associated API groups.

> == Setting Resource Quota for Extended Resources

Reviewer: Separate procedure module

> Overcommitment of resources is not allowed for extended resources, so you must specify `requests` and `limits` for the same extended resource in a quota.

Reviewer: I probably don't have an idea what overcommitment means here. Also I don't think `overcommitment` is a valid English word. Can we rewrite it to something more specific.

```suggestion
You must specify `requests` and `limits` for the same extended resource in a quota.
```

> Currently, only quota items with the prefix `requests.` are allowed for extended resources. The following is an example scenario of how to set resource quota for the GPU resource `nvidia.com/gpu`.

Reviewer: Is this an example or a procedure? I believe latter. You can specify the sample data in individual steps.
```suggestion
Currently, only quota items with the prefix `requests.` are allowed for extended resources. 
```

.Procedure

. Determine how many GPUs are available on a node in your cluster. For example:
+
[source,terminal]
----

> # oc describe node ip-172-31-27-209.us-west-2.compute.internal | egrep 'Capacity|Allocatable|gpu'

Reviewer: I think it's an oc command which should be having $ sign? Feel free to ignore if I am wrong.
```suggestion
$ oc describe node ip-172-31-27-209.us-west-2.compute.internal | egrep 'Capacity|Allocatable|gpu'
```

                    openshift.com/gpu-accelerator=true
Capacity:
 nvidia.com/gpu:  2
Allocatable:
 nvidia.com/gpu:  2

>   nvidia.com/gpu  0           0

Reviewer: is this output of command? If so,it needs to be in a separate Example output section.

Author: @snarayan-redhat Done! Thank you, I missed that one!

----
+
In this example, 2 GPUs are available.

. Set a quota in the namespace `nvidia`. In this example, the quota is `1`:
+
[source,terminal]
----

> # cat gpu-quota.yaml

Reviewer: ```suggestion
$ cat gpu-quota.yaml
```

apiVersion: v1
kind: ResourceQuota
metadata:
  name: gpu-quota
  namespace: nvidia
spec:
  hard:

>     requests.nvidia.com/gpu: 1

Reviewer: separate output section

Author: Done, thank you!

----

. Create the quota:
+
[source,terminal]
----

> # oc create -f gpu-quota.yaml

Reviewer: ```suggestion
$ oc create -f gpu-quota.yaml
```

resourcequota/gpu-quota created
----

. Verify that the namespace has the correct quota set:
+
[source,terminal]
----

> # oc describe quota gpu-quota -n nvidia

Reviewer: ```suggestion
$ oc describe quota gpu-quota -n nvidia
```

Name:                    gpu-quota
Namespace:               nvidia
Resource                 Used  Hard
--------                 ----  ----
requests.nvidia.com/gpu  0     1

> ----

Reviewer: Separate output section

Author: Done, thank  you!

> . Run a pod that asks for a single GPU:

Reviewer: ```suggestion
. Run a pod that requests for a single GPU:
```

+
[source,terminal]
----

> # oc create pod gpu-pod.yaml

Reviewer: ```suggestion
$ oc create pod gpu-pod.yaml
```

Reviewer: This comment applies to all instances of `#` .

----
+
----
apiVersion: v1
kind: Pod
metadata:
  generateName: gpu-pod-
  namespace: nvidia
spec:
  restartPolicy: OnFailure
  containers:
  - name: rhel7-gpu-pod
    image: rhel7
    env:
      - name: NVIDIA_VISIBLE_DEVICES
        value: all
      - name: NVIDIA_DRIVER_CAPABILITIES
        value: "compute,utility"
      - name: NVIDIA_REQUIRE_CUDA
        value: "cuda>=5.0"

    command: ["sleep"]
    args: ["infinity"]

    resources:
      limits:
        nvidia.com/gpu: 1

> ----

Reviewer: Separate output section

. Verify that the pod is running:
+
[source,terminal]
----
# oc get pods
NAME              READY     STATUS      RESTARTS   AGE
gpu-pod-s46h7     1/1       Running     0          1m
----

. Verify that the quota `Used` counter is correct:
+
[source,terminal]
----
# oc describe quota gpu-quota -n nvidia
Name:                    gpu-quota
Namespace:               nvidia
Resource                 Used  Hard
--------                 ----  ----
requests.nvidia.com/gpu  1     1
----

> . Attempt to create a second GPU pod in the `nvidia` namespace. This is technically available on the node because it has 2 GPUs:

Reviewer: This could be a part of Verification section, IIUC.

+
[source,terminal]
----
# oc create -f gpu-pod.yaml

> Error from server (Forbidden): error when creating "gpu-pod.yaml": pods "gpu-pod-f7z2w" is forbidden: exceeded quota: gpu-quota, requested: requests.nvidia.com/gpu=1, used: requests.nvidia.com/gpu=1, limited: requests.nvidia.com/gpu=1

Reviewer: Separate output section

Author: Thank you! Corrected this for every example in this module.

Author: There were more in the other modules. Fixing those also.

----
+
This `Forbidden` error message is expected because you have a quota of 1 GPU and this pod tried to allocate a second GPU, which exceeds its quota.

== Quota Scopes

> Each quota can have an associated set of _scopes_. A quota will only

Reviewer: ```suggestion
Each quota can have an associated set of _scopes_. A quota only
```

Author: Done, thank  you! Also repaired other issues in that same sentence.

> measure usage for a resource if it matches the intersection of enumerated

Reviewer: ```suggestion
measures usage for a resource if it matches the intersection of enumerated
```

scopes.

Adding a scope to a quota restricts the set of resources to which that quota can
apply. Specifying a resource outside of the allowed set results in a validation
error.

[cols="3a,8a",options="header"]
|===

|Scope |Description

|`Terminating`
|Match pods where `spec.activeDeadlineSeconds >= 0`.

|`NotTerminating`
|Match pods where `spec.activeDeadlineSeconds` is `nil`.

|`BestEffort`
|Match pods that have best effort quality of service for either `*cpu*` or `*memory*`.

ifdef::openshift-enterprise,openshift-origin[]

> See xref:../nodes/clusters/nodes-cluster-overcommit.adoc#nodes-cluster-overcommit-resource-requests_nodes-cluster-overcommit[Quality of Service Classes] for more on committing compute resources.

Reviewer: xrefs in additional resources

Author: Done! Great catch, thank you!

apply. Specifying a resource outside of the allowed set results in a validation
error.

[cols="3a,8a",options="header"]
|===

|Scope |Description

|`Terminating`
|Match pods where `spec.activeDeadlineSeconds >= 0`.

|`NotTerminating`
|Match pods where `spec.activeDeadlineSeconds` is `nil`.

|`BestEffort`
|Match pods that have best effort quality of service for either `*cpu*` or `*memory*`.

ifdef::openshift-enterprise,openshift-origin[]

See xref:../nodes/clusters/nodes-cluster-overcommit.adoc#nodes-cluster-overcommit-resource-requests_nodes-cluster-overcommit[Quality of Service Classes] for more on committing compute resources.

endif::openshift-enterprise,openshift-origin[]

|otBestEffort`
|Match pods that do not have best effort quality of service for `*cpu*` and `*memory*`.
|===

A `BestEffort` scope restricts a quota to limiting the following resources:

- `pods`

A `Terminating`, `NotTerminating`, and `NotBestEffort` scope restricts a quota to tracking the following resources:

- `pods`
- `memory`
- `requests.memory`
- `limits.memory`
- `cpu`
- `requests.cpu`
- `limits.cpu`
- `ephemeral-storage`
- `requests.ephemeral-storage`

> - `limits.ephemeral-storage`

Reviewer: Replace `-` with `*`

Author: Done, thank  you!

Overcommitment of resources is not allowed for extended resources, so you must specify `requests` and `limits` for the same extended resource in a quota.
Currently, only quota items with the prefix `requests.` are allowed for extended resources. The following is an example scenario of how to set resource quota for the GPU resource `nvidia.com/gpu`.

.Procedure

. Determine how many GPUs are available on a node in your cluster. For example:
+
[source,terminal]
----
$ oc describe node ip-172-31-27-209.us-west-2.compute.internal | egrep 'Capacity|Allocatable|gpu'

.Example output

                    openshift.com/gpu-accelerator=true
Capacity:
 nvidia.com/gpu:  2
Allocatable:
 nvidia.com/gpu:  2
  nvidia.com/gpu  0           0
----
+
In this example, 2 GPUs are available.

. Set a quota in the namespace `nvidia`. In this example, the quota is `1`:
+
[source,terminal]
----
$ cat gpu-quota.yaml

.Example output

apiVersion: v1
kind: ResourceQuota
metadata:
  name: gpu-quota
  namespace: nvidia
spec:
  hard:
    requests.nvidia.com/gpu: 1
----

. Create the quota:
+
[source,terminal]
----
$ oc create -f gpu-quota.yaml

.Example output

resourcequota/gpu-quota created
----

> . Verify that the namespace has the correct quota set:

Author: Done, thank you!

+
[source,terminal]
----
$ oc describe quota gpu-quota -n nvidia

.Example output

Name:                    gpu-quota
Namespace:               nvidia
Resource                 Used  Hard
--------                 ----  ----
requests.nvidia.com/gpu  0     1
----

. Run a pod that asks for a single GPU:
+
[source,terminal]
----
$ oc create pod gpu-pod.yaml
----
+

> ----

Reviewer: Output title

Author: This is done, thank you!

apiVersion: v1
kind: Pod
metadata:
  generateName: gpu-pod-
  namespace: nvidia
spec:
  restartPolicy: OnFailure
  containers:
  - name: rhel7-gpu-pod
    image: rhel7
    env:
      - name: NVIDIA_VISIBLE_DEVICES
        value: all
      - name: NVIDIA_DRIVER_CAPABILITIES
        value: "compute,utility"
      - name: NVIDIA_REQUIRE_CUDA
        value: "cuda>=5.0"

    command: ["sleep"]
    args: ["infinity"]

    resources:
      limits:
        nvidia.com/gpu: 1
----

> . Verify that the pod is running:

Reviewer: This suggestion applies to all steps containing command which misses the below piece.
```suggestion
. Verify that the pod is running by running the following command:
```

Author: Done, thank you!

+
[source,terminal]
----
$ oc get pods
----
+
.Example output
+
NAME              READY     STATUS      RESTARTS   AGE
gpu-pod-s46h7     1/1       Running     0          1m

. Verify that the quota `Used` counter is correct:
+
[source,terminal]
----
$ oc describe quota gpu-quota -n nvidia
----
+
.Example output
+
Name:                    gpu-quota
Namespace:               nvidia
Resource                 Used  Hard
--------                 ----  ----
requests.nvidia.com/gpu  1     1

. Attempt to create a second GPU pod in the `nvidia` namespace. This is technically available on the node because it has 2 GPUs:
+
[source,terminal]
----
$ oc create -f gpu-pod.yaml
----
+
.Example output
+
Error from server (Forbidden): error when creating "gpu-pod.yaml": pods "gpu-pod-f7z2w" is forbidden: exceeded quota: gpu-quota, requested: requests.nvidia.com/gpu=1, used: requests.nvidia.com/gpu=1, limited: requests.nvidia.com/gpu=1

> This `Forbidden` error message is expected because you have a quota of 1 GPU and this pod tried to allocate a second GPU, which exceeds its quota.

Reviewer: ```suggestion
This `Forbidden` error message occurs because you have a quota of 1 GPU and this pod tried to allocate a second GPU, which exceeds its quota.
```

== Quota scopes

Each quota can have an associated set of _scopes_. A quota only measures usage for a resource if it matches the intersection of enumerated scopes.

Adding a scope to a quota restricts the set of resources to which that quota can apply. Specifying a resource outside of the allowed set results in a validation error.

[cols="3a,8a",options="header"]
|===

|Scope |Description

|`Terminating`
|Match pods where `spec.activeDeadlineSeconds >= 0`.

|`NotTerminating`
|Match pods where `spec.activeDeadlineSeconds` is `nil`.

|`BestEffort`
|Match pods that have best effort quality of service for either `*cpu*` or `*memory*`.

> |otBestEffort`

Author: Great catch, thank you!

You can configure an object count quota for these standard namespaced resource types using the `count/<resource>.<group>` syntax.

[source,terminal]
----
$ oc create quota <name> --hard=count/<resource>.<group>=<quota> <1>
----

<1> `<resource>` is the name of the resource, and `<group>` is the API group, if applicable.
Use the `kubectl api-resources` command for a list of resources and their associated API groups.

== Setting resource quota for extended resources

Overcommitment of resources is not allowed for extended resources, so you must specify `requests` and `limits` for the same extended resource in a quota. Currently, only quota items with the prefix `requests.` are allowed for extended resources. The following is an example scenario of how to set resource quota for the GPU resource `nvidia.com/gpu`.

.Procedure

. To determine how many GPUs are available on a node in your cluster, use the following command:
+
[source,terminal]
----
$ oc describe node ip-172-31-27-209.us-west-2.compute.internal | egrep 'Capacity|Allocatable|gpu'
----
+
.Example output

                    openshift.com/gpu-accelerator=true
Capacity:
 nvidia.com/gpu:  2
Allocatable:
 nvidia.com/gpu:  2

>   nvidia.com/gpu  0           0

Reviewer 2: Should be align with above line?

Author: @wangke19 Looking now - I am online so I can catch your feedback while you are here!

Author: @wangke19  Thank you, I have made this change!

+
In this example, 2 GPUs are available.

. Use this command to set a quota in the namespace `nvidia`. In this example, the quota is `1`:
+
[source,terminal]
----
$ cat gpu-quota.yaml
----
+
.Example output

[source,terminal]
----
apiVersion: v1
kind: ResourceQuota
metadata:
  name: gpu-quota
  namespace: nvidia
spec:
  hard:
    requests.nvidia.com/gpu: 1

----
. Create the quota with the following command:
+
[source,terminal]
----
$ oc create -f gpu-quota.yaml
----
+
.Example output

[source,terminal]
----
resourcequota/gpu-quota created
----
. Verify that the namespace has the correct quota set using the following command:
+
[source,terminal]
----
$ oc describe quota gpu-quota -n nvidia
----
+
.Example output

[source,terminal]
----
Name:                    gpu-quota
Namespace:               nvidia
Resource                 Used  Hard
--------                 ----  ----
requests.nvidia.com/gpu  0     1
----

. Run a pod that asks for a single GPU with the following command:
+
[source,terminal]
----
$ oc create pod gpu-pod.yaml
----
+
.Example output

[source, terminal]
----
apiVersion: v1
kind: Pod
metadata:

>   generateName: gpu-pod-

Reviewer 2: Is name `gpu-pod`?

Author: Yes! Thank you! This is fixed.

Author: Yes, thank you! This is fixed!

// Module included in the following assemblies:
//
// ../scalability_and_performance/compute-resource-quotas.adoc

:_mod-docs-content-type: CONCEPT
[id="admin-quota-overview_{context}"]
= Resources managed by quota

A resource quota, defined by a `ResourceQuota` object, provides constraints that limit aggregate resource consumption per project. It can limit the quantity
of objects that can be created in a project by type, as well as the total amount of compute resources and storage that may be consumed by resources in that project.

The following describes the set of compute resources and object types that may be managed by a quota.

[NOTE]
====

> A pod is in a terminal state if `status.phase`` is `Failed` or `Succeeded`.

Author: Done, thank you!

Adding a scope to a quota restricts the set of resources to which that quota can apply. Specifying a resource outside of the allowed set results in a validation error.

[cols="3a,8a",options="header"]
|===

|Scope |Description

|`Terminating`
|Match pods where `spec.activeDeadlineSeconds >= 0`.

|`NotTerminating`
|Match pods where `spec.activeDeadlineSeconds` is `nil`.

|`BestEffort`
|Match pods that have best effort quality of service for either `cpu` or `memory`.

|`otBestEffort`
|Match pods that do not have best effort quality of service for `cpu` and `memory`.
|===

A `BestEffort` scope restricts a quota to limiting the following resources:

- `pods`

A `Terminating`, `NotTerminating`, and `NotBestEffort` scope restricts a quota to tracking the following resources:

* `pods`
* `memory`
* `requests.memory`
* `limits.memory`
* `cpu`
* `requests.cpu`
* `limits.cpu`
* `ephemeral-storage`
* `requests.ephemeral-storage`
* `limits.ephemeral-storage`

[NOTE]
====
Ephemeral storage requests and limits apply only if you enabled the ephemeral storage technology preview. This feature is disabled by default.
====

> .Additional resources

Reviewer 3: This needs to be at the assembly level.

Author: This is done, thank you!
